File: services/subtraction-service/src/main/java/com/calculator/subtraction/domain/InputValidator.java
package com.calculator.subtraction.domain;

import org.springframework.stereotype.Component;

import java.math.BigDecimal;

public interface InputValidator {
    void validate(CalculationRequest request);
}

@Component
class SubtractionInputValidator implements InputValidator {
    @Override
    public void validate(CalculationRequest request) {
        if (request.requestId() == null || request.requestId().isBlank()) {
            throw new IllegalArgumentException("Request ID cannot be blank");
        }
        
        if (request.operand1().scale() > 10 || request.operand2().scale() > 10) {
            throw new IllegalArgumentException("Operands cannot have more than 10 decimal places");
        }
        
        BigDecimal limit = new BigDecimal("1E+10");
        if (request.operand1().abs().compareTo(limit) > 0 || 
            request.operand2().abs().compareTo(limit) > 0) {
            throw new IllegalArgumentException("Operands cannot exceed 10 billion");
        }
    }
}

File: services/addition-service/src/main/kotlin/com/calculator/addition/domain/InputValidator.kt
package com.calculator.addition.domain

import java.math.BigDecimal

interface InputValidator {
    fun validate(request: CalculationRequest)
}

class AdditionInputValidator : InputValidator {
    override fun validate(request: CalculationRequest) {
        if (request.requestId.isBlank()) {
            throw IllegalArgumentException("Request ID cannot be blank")
        }
        
        if (request.operand1.scale() > 10 || request.operand2.scale() > 10) {
            throw IllegalArgumentException("Operands cannot have more than 10 decimal places")
        }
        
        if (request.operand1.abs() > BigDecimal("1E+10") || request.operand2.abs() > BigDecimal("1E+10")) {
            throw IllegalArgumentException("Operands cannot exceed 10 billion")
        }
    }
}